contracts/src/v0.8/ccip/pools/SiloedLockReleaseTokenPool.sol

```diff
@@ -0,0 +1,286 @@
+  /// non-siloed chains.
+  /// @param amount The amount of liquidity to provide.
+  /// @dev Only the rebalancer for the chain can add liquidity
+  function provideSiloedLiquidity(uint64 remoteChainSelector, uint256 amount) external {
+    _provideLiquidity(remoteChainSelector, amount);
+  }
+
+  /// @notice Adds liquidity to the pool for unsiloed chains. Function is used to support legacy liquidity operations
+  /// by using a function selector available to previous L/R pools.
+  /// @dev Since the remoteChainSelector 0 should never be applied to a real chain, it is used to designate unsiloed chains.
+  /// @param amount The amount of liquidity to provide.
+  function provideLiquidity(
+    uint256 amount
+  ) external {
+    _provideLiquidity(0, amount);
+  }
+
+  function _provideLiquidity(uint64 remoteChainSelector, uint256 amount) internal {
+    if (msg.sender != getRebalancerByChain(remoteChainSelector)) revert Unauthorized(msg.sender);
+
+    // Storage is used instead of memory to save gas, as the state may need to be updated if the chain is siloed.
+    ChainSiloConfig storage remoteConfig = s_siloedChainConfigs[remoteChainSelector];
+
+    if (remoteConfig.isSiloed) {
+      remoteConfig.siloedBalance += amount;
+    } else {
+      s_unsiloedTokenBalance += amount;
+    }
+
+    i_token.safeTransferFrom(msg.sender, address(this), amount);
+    emit LiquidityAdded(remoteChainSelector, msg.sender, amount);
+  }
+
+  /// @notice Removes liquidity from the pool for unsiloed chains. Function is used to support legacy liquidity operations
+  /// by using a function selector available to previous L/R pools.
+  /// @dev Since the remoteChainSelector 0 should never be applied to a real chain, it is used to designate unsiloed chains.
+  /// @param amount The amount of liquidity to remove.
+  function withdrawLiquidity(
+    uint256 amount
+  ) external {
+    _withdrawLiquidity(0, amount);
+  }
+
+  /// @notice Removed liquidity to the pool. The tokens will be sent to msg.sender.
+  /// @dev Only the rebalancer can remove liquidity from the contract, for both siloed and unsiloed chains.
+  /// @param remoteChainSelector the remote chain to set. If the chain is not siloed, then no accounting will be updated,
+  /// which can be considered the liquidity for all non-siloed chains sharing liquidity.
+  /// @param amount The amount of liquidity to remove.
+  function withdrawSiloedLiquidity(uint64 remoteChainSelector, uint256 amount) external {
+    _withdrawLiquidity(remoteChainSelector, amount);
+  }
+
+  function _withdrawLiquidity(uint64 remoteChainSelector, uint256 amount) internal {
+    if (msg.sender != getRebalancerByChain(remoteChainSelector)) revert Unauthorized(msg.sender);
+
+    // If funds are siloed by chain, prevent more than has been locked from being removed from the token pool.
+    if (s_siloedChainConfigs[remoteChainSelector].isSiloed) {
```

Comment:
Let's have proper error handling here as well

Reply:
The error handling is that it will underflow and revert, do you need to do an explicit check for the underflow anyways? That only adds gas.